### PR TITLE
Feat/78 session

### LINKED ISSUE #78: Session Management
- [ ] correct credentials at login page create new session
- [ ] username is written into annotation from session info
- [ ] close (401) other routes except for login page when no user is logged in

### PR DESCRIPTION
**Pull request by bewee**
### Changes made by this Pull Request

- Introduce session management via JWTs

### Definition of Done

- [x] Tests are written and run successful
- [x] Code is documented
- [x] User Manual is updated
- [x] All requirements of the user story are met

### CODE REVIEW

**Comment by bewee** [comment]
We based all our work on this tutorial: https://testdriven.io/blog/fastapi-jwt-auth/

File: api/mod/src/settings.py [hunk 1]
```diff
@@ -23,6 +23,9 @@ class Settings(BaseSettings):
         f"sqlite:///{Path.home().joinpath('.annoto').joinpath('db.sqlite3')}"
     )
     debug_routes: bool = True
+    jwt_algorithm: str = "HS256"
```
Comment (jdsauer) [inline api/mod/src/settings.py:26]: maybe we could provide some information about the HS256 in case people wants to know more about the security of the backend and session handling

File: api/mod/src/auth/auth_bearer.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+from typing import Any, Coroutine, Optional
+
+from fastapi import HTTPException, Request
+from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer
+
+from .auth_handler import decodeJWT
```
Comment (jb3rndt) [inline api/mod/src/auth/auth_bearer.py]: An absolute import would be more consistent I think

File: api/mod/src/auth/auth_bearer.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+                raise HTTPException(
+                    status_code=403, detail="Invalid authentication scheme."
+                )
+            if not self.verify_jwt(credentials.credentials):
+                raise HTTPException(
+                    status_code=403, detail="Invalid token or expired token."
+                )
+            return credentials
+        else:
+            raise HTTPException(status_code=403, detail="Invalid authorization code.")
```
Comment (jb3rndt) [inline api/mod/src/auth/auth_bearer.py]: I propose a guard clause would look a bit nicer :P
```python
if not credentials:
    raise HTTPException(status_code=403, detail="Invalid authorization code.")
if not credentials.scheme == "Bearer":
    raise HTTPException(
        status_code=403, detail="Invalid authentication scheme."
    )
if not self.verify_jwt(credentials.credentials):
    raise HTTPException(
        status_code=403, detail="Invalid token or expired token."
    )
return credentials
```

File: api/mod/src/auth/auth_bearer.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        if credentials:
+            if not credentials.scheme == "Bearer":
+                raise HTTPException(
+                    status_code=403, detail="Invalid authentication scheme."
+                )
+            if not self.verify_jwt(credentials.credentials):
+                raise HTTPException(
+                    status_code=403, detail="Invalid token or expired token."
+                )
+            return credentials
```
Comment (jb3rndt) [inline api/mod/src/auth/auth_bearer.py]: In the Tut it is `credentials.credentials` here. Why do we not want to return the jwt directly?

File: api/mod/src/auth/auth_bearer.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+from .auth_handler import decodeJWT
+
+
+class JWTBearer(HTTPBearer):
+    def __init__(self, auto_error: bool = True):
+        super(JWTBearer, self).__init__(auto_error=auto_error)
+
+    async def __call__(
+        self, request: Request
+    ) -> Optional[HTTPAuthorizationCredentials]:
```
Comment (jb3rndt) [inline api/mod/src/auth/auth_bearer.py]: In the method we check if the credentials exist, so the method cant return None, can it?

File: api/mod/src/auth/auth_bearer.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    def verify_jwt(self, jwtoken: str) -> bool:
+        isTokenValid: bool = False
+
+        try:
+            payload = decodeJWT(jwtoken)
+        except:
+            payload = None
+        if payload:
+            isTokenValid = True
+        return isTokenValid
```
Comment (jb3rndt) [inline api/mod/src/auth/auth_bearer.py]: We could save some lines here by dropping `isTokenValid` and rather just doing `return payload is not None`

File: api/mod/src/auth/auth_handler.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import time
+from typing import Dict, Optional
+
+import jwt
+
+from mod.src.settings import SETTINGS
+
+
+class Payload:
```
Comment (jb3rndt) [inline api/mod/src/auth/auth_handler.py]: When I initially read `Payload` in the other methods, I thought it is a class from fastapi that contains the whole request payload. What do you think of renaming it to something like `JWTPayload`?

File: api/mod/src/auth/auth_bearer.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+                )
+            return credentials
+        else:
+            raise HTTPException(status_code=403, detail="Invalid authorization code.")
+
+    def verify_jwt(self, jwtoken: str) -> bool:
+        isTokenValid: bool = False
+
+        try:
+            payload = decodeJWT(jwtoken)
```
Comment (jb3rndt) [inline api/mod/src/auth/auth_bearer.py]: `decodeJWT` already catches all exceptions, so in theory we dont need the try except here. Or do we want to keep that for safety?

File: api/mod/src/routers/images.py [hunk 9]
```diff
@@ -78,19 +86,30 @@ async def get_image(
...
     annotation_data: AnnotationData,
     task_id: str = Path(..., example="ecg-qrs-classification-physiodb"),
     src: str = Path(..., example="sloth.jpg"),
+    jwt: Optional[HTTPAuthorizationCredentials] = Depends(JWTBearer()),
+    db: Session = Depends(DB),
 ) -> None:
     """Saves the annotation for the specified image"""
 
+    if not jwt:
+        raise HTTPException(status_code=500)
```
Comment (jb3rndt) [inline api/mod/src/routers/images.py:103]: If I understand correctly, the error for not being authorized should already be thrown in the `JWTBearer.__call__` method. Is that why this is considered an internal server error here? Will that even ever be triggered?

File: api/mod/src/routers/images.py [hunk 10]
```diff
@@ -78,19 +86,30 @@ async def get_image(
...
+
+    user_id = decodeJWT(jwt.credentials).user_id
```
Comment (jb3rndt) [inline api/mod/src/routers/images.py]: should we handle that this can return None?

File: api/mod/src/settings.py [hunk 11]
```diff
@@ -23,6 +23,9 @@ class Settings(BaseSettings):
...
+    jwt_secret: str = "fd232ad1ff70a3a874dda48e857b1ef086e474194ce5d2f8"
```
Comment (jb3rndt) [inline api/mod/src/settings.py:27]: I dont exactly know how to handle this secret, but it should not be exposed in github/git, should it?

File: api/mod/tests/integration/conftest.py [hunk 12]
```diff
@@ -3,12 +3,20 @@
...
 @pytest.fixture
 def user() -> CreateUserRequest:
     """Creates a basic Annotation"""
     return CreateUserRequest(
-        fullname="Prof. Dr. Folivora",
-        password="password",
-        email="team@folivora.online",
+        fullname="Prof. Dr. Perry",
+        password="brrrrrrr",
+        email="perry@folivora.online",
```
Comment (jb3rndt) [inline api/mod/tests/integration/conftest.py]: Wyld =D
Will be difficult to remember exactly how many "r" will be needed in the pw when writing tests :P

File: api/mod/tests/integration/test_images_router.py [hunk 13]
```diff
@@ -9,21 +9,27 @@
...
 
 
-def test_get_image(client: TestClient) -> None:
+def test_get_image(client: TestClient, authorization: str) -> None:
     """Test GET /tasks/ecg-qrs-classification-physiodb/sloth.jpg"""
-    response = client.get("/tasks/ecg-qrs-classification-physiodb/sloth.jpg")
+    response = client.get(
+        "/tasks/ecg-qrs-classification-physiodb/sloth.jpg",
+        headers={"Authorization": authorization},
+    )
```
Comment (jb3rndt) [inline api/mod/tests/integration/test_images_router.py:17]: So we'll need to add that for every test (that requests a route that needs authorization) I guess. Do you think there is a way to create a fixture like `authorized_client` that is a tweaked `TestClient` that automatically sends requests with Authorization header?

File: api/mod/tests/integration/test_images_router.py [hunk 14]
```diff
@@ -32,9 +38,10 @@ def test_post_image(client: TestClient) -> None:
             "hash": "e922903b4d5431a8f9def3c89ffcb0b18472f3da304f28a2dbef9028b6cd205d",
             "competency": "Prof. Dr. Med",
             "is_attentive": True,
-            "username": "Prof. Dr. Folivora",
+            "fullname": "Prof. Dr. Folivora",
```
Comment (jb3rndt) [inline api/mod/tests/integration/test_images_router.py]: Funny that fastapi throws no error, but the fullname key is not included in the `AnnotationData`, right? Also applies to some other tests

File: api/mod/tests/integration/test_login_router.py [hunk 15]
```diff
@@ -28,13 +24,42 @@ def test_unknown_user_returns_401(client: TestClient) -> None:
...
+    )
+    assert response.status_code == 200
+    token = response.json()["access_token"]
+    response = client.get("/tasks", headers={"Authorization": f"Bearer {token}"})
+    assert response.status_code == 200
+
+
+def test_login_token_expires(client: TestClient, mocker: MockerFixture) -> None:
+    """Test POST /login"""
+    mocker.patch("time.time", lambda: 0)
```
Comment (jb3rndt) [inline api/mod/tests/integration/test_login_router.py]: I think you dont need a lambda here. `return_value=0` works too. What do you prefer? For consistency we should change it in other places as well, if we want to keep lambda

File: api/mod/tests/integration/test_users_router.py [hunk 16]
```diff
@@ -13,8 +13,8 @@ def test_get_user_returns_correct_user(
     client: TestClient, db: Session, user: CreateUserRequest
 ) -> None:
     """Test GET /users/1"""
-    DBUser.create(db=db, user=user)
-    response = client.get("/users/1")
+    user = DBUser.create(db=db, user=user)
```
Comment (jb3rndt) [inline api/mod/tests/integration/test_users_router.py]: I think user creation can be removed since it is done in the `db_engine` fixture now

File: static/src/components/AnnotationButton.vue [hunk 17]
```diff
@@ -43,7 +43,6 @@ export default defineComponent({
         await API.saveAnnotation(this.taskId, this.imageId, {
           label: this.label,
           hash: hash,
-          username: this.username,
```
Comment (jb3rndt) [inline static/src/components/AnnotationButton.vue:46]: Can we remove the username prop now too?

File: static/src/components/ImageDisplay.vue [hunk 18]
```diff
@@ -16,16 +18,25 @@ export default defineComponent({
...
+    };
+  },
+
+  mounted() {
+    this.src = [
+      computedAsync(async () => {
+        let image = await API.getImage(this.taskId, this.imageId);
+        return image ? URL.createObjectURL(image) : undefined;
+      }, undefined),
+    ];
```
Comment (jb3rndt) [inline static/src/components/ImageDisplay.vue:33]: Sorry I forgot, but this returns the image so we dont have to use the url in the `` tag which does not work due to authorization right? Nice that you were able to integrate that.
A possible solution for the async thing might be vue asyncComponents. But im fine if we put that in a different pr

**Review by jb3rndt — COMMENTED** [review]
Nice work 👌
Im still trying to get into the topic, so some comments might be useless

File: api/mod/src/auth/auth_bearer.py (hunk 4 above)
Comment (bewee) [reply api/mod/src/auth/auth_bearer.py]: I did not figure out how else to get the typing correctly :see_no_evil:

File: api/mod/src/routers/images.py (hunk 10 above)
Comment (bewee) [reply api/mod/src/routers/images.py]: Wow, I'd expect mypy to complain about just these things :thinking:

File: api/mod/src/settings.py (hunk 11 above)
Comment (bewee) [reply api/mod/src/settings.py:27]: Yes and no. I think it's ok for our testing/dev system. We should however definitely overwrite it in the `.env.production` file

File: api/mod/tests/integration/conftest.py (hunk 12 above)
Comment (bewee) [reply api/mod/tests/integration/conftest.py]: I may have been quite tired already at this point, so.... well.... I'd say: Open for change :sweat_smile:

File: api/mod/tests/integration/test_images_router.py (hunk 13 above)
Comment (bewee) [reply api/mod/tests/integration/test_images_router.py:17]: Neat idea. Don't know whether / how that's possible, but would definitely be a great idea!

File: api/mod/tests/integration/test_images_router.py (hunk 14 above)
Comment (bewee) [reply api/mod/tests/integration/test_images_router.py]: Presumably because everything it needs is present, and the remainder will just be ignored :shrug:

File: api/mod/tests/integration/test_login_router.py (hunk 15 above)
Comment (bewee) [reply api/mod/tests/integration/test_login_router.py]: return_value=0 is perfect

File: static/src/components/AnnotationButton.vue (hunk 17 above)
Comment (bewee) [reply static/src/components/AnnotationButton.vue:46]: Yeah, but I'd suggest moving this to the other front-end related username/fullname PR - What do you think?

File: static/src/components/ImageDisplay.vue (hunk 18 above)
Comment (bewee) [reply static/src/components/ImageDisplay.vue:33]: Let's move it to a separate issue, there's already one: #62

File: api/mod/src/settings.py (hunk 1 above)
Comment (bewee) [reply api/mod/src/settings.py:26]: Actually, I don't know anything more about it either... So if you have something to add, feel free to just push it :)

File: api/mod/src/settings.py (hunk 11 above)
Comment (bewee) [reply api/mod/src/settings.py:27]: * or using env variables

File: api/mod/src/auth/auth_bearer.py (hunk 4 above)
Comment (jb3rndt) [reply api/mod/src/auth/auth_bearer.py]: Kk, me neither

File: api/mod/src/settings.py (hunk 11 above)
Comment (jb3rndt) [reply api/mod/src/settings.py:27]: Yep, should we consider that for staging too?

File: static/src/components/AnnotationButton.vue (hunk 17 above)
Comment (jb3rndt) [reply static/src/components/AnnotationButton.vue:46]: Ok, I referenced it in #95

File: api/mod/tests/integration/test_images_router.py (hunk 13 above)
Comment (jb3rndt) [reply api/mod/tests/integration/test_images_router.py:17]: Ok, but lets put it in another pr

**Review by jb3rndt — APPROVED** [review]
The pending conversations are far from being mandatory to solve from my point of view, so feel free to merge :)

File: api/mod/src/settings.py (hunk 11 above)
Comment (bewee) [reply api/mod/src/settings.py:27]: Sounds reasonable

**Comment by bewee** [comment]
Let's do it then!
